src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint_sd_xl.py

```diff
@@ -867,8 +867,20 @@ def get_timesteps(self, num_inference_steps, strength, device, denoising_start=N
                     - (denoising_start * self.scheduler.config.num_train_timesteps)
                 )
             )
-            timesteps = list(filter(lambda ts: ts < discrete_timestep_cutoff, timesteps))
-            return torch.tensor(timesteps), len(timesteps)
+
+            num_inference_steps = (timesteps < discrete_timestep_cutoff).sum().item()
+            if self.scheduler.order == 2:
+                # if the scheduler is a 2nd order scheduler we ALWAYS have to do +1
+                # because `num_inference_steps` will always be even given that every timestep
+                # (except the highest one) is duplicated. If `num_inference_steps` is even it would
+                # mean that we cut the timesteps in the middle of the denoising step
+                # (between 1st and 2nd devirative) which leads to incorrect results. By adding 1
+                # we ensure that the denoising process always ends after the 2nd derivate step of the scheduler
+                num_inference_steps = num_inference_steps + 1
```

Comment:
I don't think `num_inference_steps` will always be even. Consider the case where the scheduler has been set up by calling `set_timesteps(25)` and then supplied `denoising_start = 0.6`. The following code prints an even number:

```python3
sched = SCHEDULERS["DPM2"]
sched.set_timesteps(25)
denoising_start = 0.6
discrete_timestep_cutoff = int(
    round(
        sched.config.num_train_timesteps
        - (denoising_start * sched.config.num_train_timesteps)
    )
)
num_inference_steps = (sched.timesteps < discrete_timestep_cutoff).sum().item()
print(num_inference_steps)
```

This prints 20. For denoising_start=0.5, it prints 25. So `num_inference_steps` may be either odd or even depending on `num_inference_steps` and `denoising_start` passed into the pipeline call.

I think we should add 1 to it only if it's even, and not if it's odd. No?

@yiyixuxu @patrickvonplaten

Reply:
Ah DPM2 can actually be different indeed if it interpolates, (Heun can't)

Nice catch! I'll open a PR to fix it

Reply:
This is the result I get with hnf = 0.5 and steps = 25 with DPM2 with the code from this PR: 
![image](https://github.com/huggingface/diffusers/assets/485488/880fc2e8-6200-4ba0-af0f-bb5f241f92c5)

Reply:
https://github.com/huggingface/diffusers/pull/5526#issue-1961250650

Reply:
Thanks @patrickvonplaten, all testcases I threw at it now produce good images.

I am just curious why we don't need to do any logic in the base step (where denoising_end is supplied to the base SDXL pipeline without img2img) since I would expect that we would want to ensure that the final timestep there is a second order timestep, and as far as I can tell it still has the odd/even issue when splitting. Maybe I am wrong!

In any case, latest `main` now produces good images for all my testcases like I said. Thank you! 🙏